client/src/main/java/io/cloudsoft/winrm4j/client/ShellCommand.java

```diff
@@ -0,0 +1,212 @@
+
+        String commandId = cmdResponse.getCommandId();
+
+        try {
+            return receiveCommand(commandId, out, err);
+        } finally {
+            try {
+                releaseCommand(commandId);
+            } catch (SOAPFaultException soapFault) {
+                assertFaultCode(soapFault, WSMAN_FAULT_CODE_SHELL_WAS_NOT_FOUND);
+            }
+        }
+    }
+
+    private int receiveCommand(String commandId, Writer out, Writer err) {
+        while(true) {
+            final Receive receive = new Receive();
+            DesiredStreamType stream = new DesiredStreamType();
+            stream.setCommandId(commandId);
+            stream.setValue("stdout stderr");
+            receive.setDesiredStream(stream);
+
+
+            try {
+                numberOfReceiveCalls++;
+                ReceiveResponse receiveResponse = winrm.receive(receive, WinRmClient.RESOURCE_URI, WinRmClient.MAX_ENVELOPER_SIZE, operationTimeout, locale, shellSelector);
+                getStreams(receiveResponse, out, err);
+
+                CommandStateType state = receiveResponse.getCommandState();
+                if (COMMAND_STATE_DONE.equals(state.getState())) {
+                    return state.getExitCode().intValue();
+                } else {
+                    LOG.debug("{} is not done. Response it received: {}", this, receiveResponse);
+                }
+            } catch (SOAPFaultException soapFault) {
+                /**
+                 * If such Exception which has a code 2150858793 the client is expected to again trigger immediately a receive request.
+                 * https://msdn.microsoft.com/en-us/library/cc251676.aspx
+                 */
+                assertFaultCode(soapFault, WSMAN_FAULT_CODE_OPERATION_TIMEOUT_EXPIRED);
+            }
+        }
+    }
+
+    private void assertFaultCode(SOAPFaultException soapFault, String code) {
+        try {
+            NodeList faultDetails = soapFault.getFault().getDetail().getChildNodes();
+            for (int i = 0; i < faultDetails.getLength(); i++) {
+                if (faultDetails.item(i).getLocalName().equals("WSManFault")) {
+                    if (faultDetails.item(i).getAttributes().getNamedItem("Code").getNodeValue().equals(code)) {
+                        LOG.trace("winrm client {} received error 500 response with code {}, response {}", this, code, soapFault);
+                        return;
+                    } else {
+                        throw soapFault;
+                    }
+                }
+            }
+            throw soapFault;
+        } catch (NullPointerException e) {
+            LOG.debug("Error reading Fault Code {}", soapFault.getFault());
+            throw soapFault;
+        }
+    }
+
+    public int getNumberOfReceiveCalls() {
+        return numberOfReceiveCalls;
```

Comment:
@bostko fixed this and verified all tests are successful :) (this time after doing all the changes!)
re deprecating `getNumberOfReceiveCalls` - this was my thinking as well, it's an implementation detail, exposed because the tests reference it (and they use it in a heuristic check). Let's keep it deprecated this release so potential users will be notified it's going away and make it accessible only to tests in future. Or figure out an alternative way to verify the behaviour without needing to expose it.